tracer/src/Datadog.Trace/Activity/ActivityLinkConverter.cs

```diff
@@ -0,0 +1,125 @@
+﻿// <copyright file="ActivityLinkConverter.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Datadog.Trace.Activity.DuckTypes;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+using Datadog.Trace.Vendors.Newtonsoft.Json.Linq;
+
+namespace Datadog.Trace.Activity;
+
+internal class ActivityLinkConverter : JsonConverter<IActivityLink>
+{
+    public override bool CanRead => false; // don't need to read the JSON only write
+
+    public override void WriteJson(JsonWriter writer, IActivityLink value, JsonSerializer serializer)
+    {
+        int numInvalidAttr = 0;
+        var linkJObject = new JObject();
+        linkJObject.Add("span_id", value.Context.SpanId.SpanId);
+        linkJObject.Add("trace_id", value.Context.TraceId.TraceId);
+
+        // allowed types as values: string, bool, or some numeric type
+        // note that char wasn't listed in RFC, but putting it here as it is like a string
+        // allow any dimension of those primitive types above
+        if (value.Tags is not null)
+        {
+            var acceptedAttr = new List<KeyValuePair<string, object>>();
+            // while we _can_ serialize most objects we aren't supposed to
+            foreach (var kvp in value.Tags)
+            {
+                if (string.IsNullOrEmpty(kvp.Key))
+                {
+                    numInvalidAttr++;
+                }
+                else
+                {
+                    if (IsAllowedType(kvp.Value))
+                    {
+                        acceptedAttr.Add(kvp);
+                    }
+                    else
+                    {
+                        numInvalidAttr++;
+                    }
+                }
+            }
+
+            if (acceptedAttr.Count > 0)
+            {
+                var jObject = new JObject(acceptedAttr.Select(kvp => new JProperty(kvp.Key, JToken.FromObject(kvp.Value))));
```

Comment:
One thing that would be nice to do is to ensure that when we receive an array that we do not encode `null` values (e.g. `attr1=["value1",null,"value3"]`). According to the spec, what we're doing right now where we keep it as-is, but ideally we wouldn't transmit them.

For this PR, can you add an example where a `null` value is set as a span tag so we can see the status of this in our snapshots?

Reply:
Closing this thread as we no longer need this class